src/System.Console/src/System/ConsolePal.Unix.cs

```diff
@@ -356,69 +362,171 @@ private static unsafe void GetCursorPosition(out int left, out int top)
+                    // that last case is very challenging, and rare, so we don't try, but we do need to
+                    // handle the rest.  The min bar here is doing something reasonable, which may include
+                    // giving up and just returning default top and left values.
+
+                    // Consume from stdin until we find all of the key markers for the CPR:
+                    // \ESC, '[', ';', and 'R'.  For everything before the \ESC, it's definitely
+                    // not part of the CPR sequence, so we just immediately move any such bytes
+                    // over to the StdInReader's extra buffer.  From there until the end, we buffer
+                    // everything into readBytes for subsequent parsing.
+                    const byte Esc = 0x1B;
                     StdInReader r = StdInReader.Inner;
-                    byte b;
+                    int escPos, bracketPos, semiPos, rPos;
+                    if (!AppendToStdInReaderUntil(Esc, r, readBytes, ref readBytesPos, out escPos) ||
+                        !BufferUntil((byte)'[', r, ref readBytes, ref readBytesPos, out bracketPos) ||
+                        !BufferUntil((byte)';', r, ref readBytes, ref readBytesPos, out semiPos) ||
+                        !BufferUntil((byte)'R', r, ref readBytes, ref readBytesPos, out rPos))
+                    {
+                        // We were unable to read everything from stdin, e.g. a timeout ocurred.
+                        // Since we couldn't get the complete CPR, transfer any bytes we did read
+                        // back to the StdInReader's extra buffer, treating it all as user input,
+                        // and exit having not computed a valid cursor position.
+                        TransferBytes(readBytes.Slice(readBytesPos), r);
+                        return;
+                    }
+
+                    // At this point, readBytes starts with \ESC and ends with 'R'.
+                    Debug.Assert(readBytesPos > 0 && readBytesPos <= readBytes.Length);
+                    Debug.Assert(escPos == 0 && bracketPos > escPos && semiPos > bracketPos && rPos > semiPos);
+                    Debug.Assert(readBytes[escPos] == Esc);
+                    Debug.Assert(readBytes[bracketPos] == '[');
+                    Debug.Assert(readBytes[semiPos] == ';');
+                    Debug.Assert(readBytes[rPos] == 'R');
+
+                    // There are other sequences that begin with \ESC + '[' and that might be in our sequence before
+                    // the CPR, so we don't immediately trust escPos and bracketPos.  Instead, as a heuristic we trust
+                    // semiPos (which we only tracked after seeing a '[' after seeing an \ESC) and search backwards from
+                    // there looking for '[' and then \ESC.
+                    bracketPos = readBytes.Slice(0, semiPos).LastIndexOf((byte)'[');
```

Comment:
Are there any other keys that are interpreted as `\ESC + [ + ... + ;`? I'm wondering if we can be confident that the smicolon will always belong the the CPR.

Reply:
> I'm wondering if we can be confident that the smicolon will always belong the the CPR.

We can't. For example, someone could type a left arrow followed by ';' before getting the cursor position. But I don't know of a good solution to that.  This is all a heuristic to try to do the right thing in the common cases and to avoid blowing up in others, but not necessarily always get 100% the right answer.  If there are ways you have in mind to improve it, I'd be interested... I landed here after exploring a few various approaches, and this seemed to strike the right balance.

Reply:
I see that makes sense, I guess that because of the mechanism we use to get the position and because the user can keep typing whatever he/she wants, then this method will always be error-prone. Thanks a lot for explaining!

Reply:
From my console output, I saw CPR code has double "[[".  e.g,   ^[[24;12R ?

Reply:
What do you mean by "my console output"?  The CPR does not contain two brackets.

Reply:
On Windows PowerShell, I run
1..5 | foreach { [console]::Write([char] 0x1b + '[6n') }

Output: ^[[5;1R^[[5;1R^[[5;1R^[[5;1R^[[5;1R

Reply:
> Output: ^[[5;1R^[[5;1R^[[5;1R^[[5;1R^[[5;1R

You're seeing the terminal's visual representation of ESC, which it renders as `^[`.

Try reading stdin... you'll see there's only one `[`.

Reply:
thanks.